### PR TITLE
BigQuery FTE

### PR DESCRIPTION
**Pull request by mwd410**
## Description

This PR introduces FTE for BigQuery.

This differs slightly from previous FTE PRs in that the FailureRecoveryTests are executed within the `-Pcloud-tests`, rather than a separate profile like the others. This _may_  cause problems by running up against the 1 hour time limit, in which case, we'll have to separate the FailureRecoveryTests from cloud-tests.

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
(X) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by arhimondr** [comment]
/test-with-secrets sha=a0db7441684de62e56d8e9a98a6d7428912fd1d7

**Comment by nineinchnick** [comment]
Here's the latest workflow run with secrets: https://github.com/trinodb/trino/actions/runs/3857770553

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java [hunk 1]
```diff
@@ -484,27 +537,99 @@ public void truncateTable(ConnectorSession session, ConnectorTableHandle tableHa
...
             columnNames.add(column.getName());
             columnTypes.add(column.getTrinoType());
         }
-        return new BigQueryInsertTableHandle(table.asPlainTable().getRemoteTableName(), columnNames.build(), columnTypes.build());
+        ColumnMetadata pageSinkIdColumn = getPageSinkIdColumn(columnNames.build());
+        tempFields.add(toField(pageSinkIdColumn.getName(), pageSinkIdColumn.getType(), pageSinkIdColumn.getComment()));
+
+        BigQueryClient client = bigQueryClientFactory.create(session);
+        String projectId = ((BigQueryNamedRelationHandle) table.getRelationHandle()).getRemoteTableName().getProjectId();
+        String schemaName = ((BigQueryNamedRelationHandle) table.getRelationHandle()).getSchemaTableName().getSchemaName();
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:559]: Why `projectId` is retrieved via `getRemoteTableName` and `schemaName` is retrieved via `getSchemaTableName` (why not `RemoteTableName#getDatasetName`)?

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java [hunk 2]
```diff
@@ -677,4 +802,18 @@ public RecordCursor cursor(ConnectorTransactionHandle transactionHandle, Connect
             }
         };
     }
+
+    private static ColumnMetadata getPageSinkIdColumn(List<String> otherColumnNames)
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:806]: nit: I would rename to `buildPageSinkIdColumn`.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java [hunk 3]
```diff
@@ -438,31 +459,63 @@ private BigQueryOutputTableHandle createTable(ConnectorSession session, Connecto
...
+        if (isWildcardTable(TableDefinition.Type.valueOf(handle.getType()), handle.getRemoteTableName().getTableName())) {
             throw new TrinoException(BIGQUERY_UNSUPPORTED_OPERATION, "This connector does not support dropping wildcard tables");
         }
-        TableId tableId = bigQueryTable.asPlainTable().getRemoteTableName().toTableId();
+        dropTable(session, handle.getRemoteTableName().toTableId());
+    }
+
+    // This is a last resort method - use one of the ones above if you have that type.
+    // Only use this if you are certain the TableId does not refer to a wildcard table.
+    private void dropTable(ConnectorSession session, TableId tableId)
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:516]: Calling the factory method in closers is simpler than separating `dropTable` method in my opinion:
```java
closer.register(() -> bigQueryClientFactory.create(session).dropTable(tempTable.toTableId()));
...
closer.register(() -> bigQueryClientFactory.create(session).dropTable(pageSinkTable.toTableId()));
```

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java (hunk 1 above)
Comment (mwd410) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:559]: i think this was a mistake

**Comment by losipiuk** [comment]
/test-with-secrets sha=ae199d10b18460e09a7989b2e1e4097b274a1982

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java [hunk 4]
```diff
@@ -484,27 +517,99 @@ public void truncateTable(ConnectorSession session, ConnectorTableHandle tableHa
...
+        tempFields.add(toField(pageSinkIdColumn.getName(), pageSinkIdColumn.getType(), pageSinkIdColumn.getComment()));
+
+        BigQueryClient client = bigQueryClientFactory.create(session);
+        String projectId = ((BigQueryNamedRelationHandle) table.getRelationHandle()).getRemoteTableName().getProjectId();
+        String schemaName = ((BigQueryNamedRelationHandle) table.getRelationHandle()).getRemoteTableName().getDatasetName();
+
+        if (!schemaExists(session, schemaName)) {
+            throw new SchemaNotFoundException(schemaName);
+        }
+        String remoteSchemaName = getRemoteSchemaName(client, projectId, schemaName);
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:544]: `schemaName` is already remote schema name. Calling `getRemoteSchemaName` method is redundant.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java [hunk 5]
```diff
@@ -438,20 +450,41 @@ private BigQueryOutputTableHandle createTable(ConnectorSession session, Connecto
...
+
+        return new BigQueryOutputTableHandle(
+                new RemoteTableName(tableId),
+                columnsNames.build(),
+                columnsTypes.build(),
+                temporaryTableName,
+                pageSinkIdColumn.map(ColumnMetadata::getName));
+    }
+
+    private TableId createTable(BigQueryClient client, String projectId, String remoteSchemaName, String tableName, ImmutableList<Field> fields, Optional<String> tableComment)
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:470]:
```suggestion
    private TableId createTable(BigQueryClient client, String projectId, String datasetName, String tableName, List<Field> fields, Optional<String> tableComment)
```

**Review by ebyhr — APPROVED** [review]
I would leave the final review & merge to @arhimondr @losipiuk

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java [hunk 6]
```diff
@@ -484,27 +517,99 @@ public void truncateTable(ConnectorSession session, ConnectorTableHandle tableHa
...
             columnNames.add(column.getName());
             columnTypes.add(column.getTrinoType());
         }
-        return new BigQueryInsertTableHandle(table.asPlainTable().getRemoteTableName(), columnNames.build(), columnTypes.build());
+        ColumnMetadata pageSinkIdColumn = buildPageSinkIdColumn(columnNames.build());
+        tempFields.add(toField(pageSinkIdColumn.getName(), pageSinkIdColumn.getType(), pageSinkIdColumn.getComment()));
+
+        BigQueryClient client = bigQueryClientFactory.create(session);
+        String projectId = ((BigQueryNamedRelationHandle) table.getRelationHandle()).getRemoteTableName().getProjectId();
+        String schemaName = ((BigQueryNamedRelationHandle) table.getRelationHandle()).getRemoteTableName().getDatasetName();
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:539]: `BigQueryTableHandle#asPlainTable` returns `BigQueryNamedRelationHandle`.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryUtil.java [hunk 7]
```diff
@@ -82,4 +83,9 @@ public static String quote(String name)
     {
         return QUOTE + name.replace(QUOTE, ESCAPED_QUOTE) + QUOTE;
     }
+
+    public static String quoted(RemoteTableName table)
```
Comment (losipiuk) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryUtil.java:87]: nit: `quote` above vs `quoted` here looks kinda unnatural. I would keep name the same.

**Comment by losipiuk** [comment]
/test-with-secrets sha=6c5eb9f63905c55b99fc49e561c1cc95579063f6

**Review by losipiuk — APPROVED** [review]

**Comment by losipiuk** [comment]
/test-with-secrets sha=6fa0767cc2d96008f2876bfa307693c4a267d79f

**Comment by losipiuk** [comment]
> `/test-with-secrets sha=6fa0767cc2d96008f2876bfa307693c4a267d79f`

run in progress:https://github.com/trinodb/trino/actions/runs/3931242353

**Comment by losipiuk** [comment]
/test-with-secrets sha=d8a1e6e3a29aac958bd95ec5ce3f950d7e7c1cc4

**Comment by losipiuk** [comment]
/test-with-secrets sha= 62b32d8cce2db86e290339d2c851b90985389009

**Comment by losipiuk** [comment]
/test-with-secrets sha=62b32d8cce2db86e290339d2c851b90985389009

**Comment by losipiuk** [comment]
/test-with-secrets sha=d53a55847e0d026a9589374e773b5770ddc8a12a

**Comment by losipiuk** [comment]
cancelled local CI as I am running version "with secrets". And we care about output of the latter.

**Comment by losipiuk** [comment]
ci run: https://github.com/trinodb/trino/actions/runs/3967428518/jobs/6799377521

**Comment by losipiuk** [comment]
/test-with-secrets sha=d53a55847e0d026a9589374e773b5770ddc8a12a

**Comment by losipiuk** [comment]
I got https://github.com/trinodb/trino/issues/15801 but it looks totally unrelated (it happens in different test job). merging.

**Pull request merged**